document/js-api/index.bs

```diff
@@ -1115,6 +1147,111 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
+            1. Perform [=EvaluateCall=] (|promise|.\[[Reject]],|result|.\[[Value]],false).
+            1. Return UNDETERMINED
+        1. Assert fatal error
+    1. Let |con| be [=CreateBuiltinFunction=](|runner|,3,"",[])
+    1. Perform [$Call$](|con|, undefined, [])
+    1. Returns |promise|
+</div>
+
+Note: The extra |$Call$| in the above algorithm ensures that the creation of the Promise is separated from the fullfilling of that Promise. In effect, this allows suspension of the fullfillment to occur whilst allowing the creation of the Promise itself to continue.
+
+<div algorithm>
+    The <dfn constructor for="Suspending">Suspending(|jsFun|)</dfn> constructor, when invoked, performs the following steps:
+    1. If [$IsCallable$](|jsFun|) is false, throw a {{TypeError}}.
+    1. Let |suspendingProto| be \[[%WebAssembly.Suspending.prototype%]]
+    1. Let |susp| be the result of [$OrdinaryObjectCreate$](|suspendingProto|)
+    1. Assign the \[[wrappedFunction]] internal slot of |susp| to |jsFun|
+    1. Return |susp|
+</div>
+
+<div algorithm>
+To <dfn>create a suspending function</dfn> from a JavaScript function |func|, with type |functype| perform the following steps:
+
+    1. Assert: [$IsCallable$](|func|).
+    1. Let |stored settings| be the <a spec=HTML>incumbent settings object</a>.
+    1. Let |hostfunc| be a [=host function=] which performs the following steps when called with arguments |arguments|:
+        1. Let |realm| be |func|'s [=associated Realm=].
+        1. Let |relevant settings| be |realm|'s [=realm/settings object=].
+        1. Let |async_context| be the [=surrounding agent=]'s [=running execution context=].
+        1. [=Prepare to run script=] with |relevant settings|.
+        1. [=Prepare to run a callback=] with |stored settings|.
+        1. Let [|parameters|] → [|results|] be |functype|.
+        1. Let |jsArguments| be the result of [=coerce WebAssembly arguments=](|parameters|,|arguments|)
+        1. Let |ret| be [=?=] [$Call$](|func|, undefined, |jsArguments|).
+        1. [=Clean up after running a callback=] with |stored settings|.
+        1. [=Clean up after running script=] with |relevant settings|.
+        1. Assert: |ret|.\[[Type]] is <emu-const>throw</emu-const> or <emu-const>normal</emu-const>.
+        1. If |ret|.\[[Type]] is <emu-const>throw</emu-const>, then trigger a WebAssembly trap, and propagate |ret|.\[[Value]] to the enclosing JavaScript.
+        1. If [=list/size=] of |ret| is 1 and [$IsPromise$](|ret|.\[[Value]][0]):
+            1. Let |promise| be |ret|.\[[Value]][0].
+            1. Suspend |async_context|, removing it from the ambient [=Execution Context Stack=].
+            1. If |async_context| is not performing the steps in [=run a Promising function=] then trigger a WebAssembly trap.
+            1. Let |fulled| be an [=AbstractClosure=] with parameters (|v|) that captures |functype|, |async_context| and performs the following steps when called:
```

Comment:
Maybe call this `|resume|` or something? `|fulled|` doesn't make much sense to me.